src/TriggerBinding/SqlTriggerListener.cs

```diff
@@ -464,6 +471,110 @@ PRIMARY KEY ({primaryKeys})
             }
         }
 
+        async Task<ScaleMetrics> IScaleMonitor.GetMetricsAsync()
+        {
+            return await this.GetMetricsAsync();
+        }
+
+        public async Task<SqlTriggerMetrics> GetMetricsAsync()
+        {
+            Debug.Assert(!(this._changeMonitor is null));
+
+            return new SqlTriggerMetrics
+            {
+                UnprocessedChangeCount = await this._changeMonitor.GetUnprocessedChangeCountAsync(),
+                Timestamp = DateTime.UtcNow,
+            };
+        }
+
+        ScaleStatus IScaleMonitor.GetScaleStatus(ScaleStatusContext context)
+        {
+            return this.GetScaleStatusCore(context.WorkerCount, context.Metrics?.Cast<SqlTriggerMetrics>().ToArray());
+        }
+
+        public ScaleStatus GetScaleStatus(ScaleStatusContext<SqlTriggerMetrics> context)
+        {
+            return this.GetScaleStatusCore(context.WorkerCount, context.Metrics?.ToArray());
+        }
+
+        private ScaleStatus GetScaleStatusCore(int workerCount, SqlTriggerMetrics[] metrics)
+        {
+            const int minMetricSamplesForScaling = 5;
+            const int maxChangesPerWorker = 1000;
+
+            var status = new ScaleStatus
+            {
+                Vote = ScaleVote.None
+            };
+
+            // Do not make a scale decision unless we have enough samples.
+            if (metrics == null || (metrics.Length < minMetricSamplesForScaling))
+            {
+                return status;
+            }
+
+            // Add worker is count of unprocessed changes per worker exceeds the maximum limit.
+            long lastUnprocessedChangeCount = metrics.Last().UnprocessedChangeCount;
+            this._logger.LogInformation($"Unprocessed change count: {lastUnprocessedChangeCount}, worker count: {workerCount}, maximum changes per worker: {maxChangesPerWorker}.");
+
+            if (lastUnprocessedChangeCount > workerCount * maxChangesPerWorker)
+            {
+                status.Vote = ScaleVote.ScaleOut;
+                this._logger.LogInformation($"Requesting scale-out: Found too many unprocessed changes for table: '{this._userTable.FullName}' relative to the number of workers.");
+                return status;
+            }
+
+            // Check to see if there are no changes occurring in the user table for a while.
+            if (metrics.All(p => p.UnprocessedChangeCount == 0L))
+            {
+                status.Vote = ScaleVote.ScaleIn;
+                this._logger.LogInformation($"Requesting scale-in: Found table: '{this._userTable.FullName}' to be idle for a while.");
+                return status;
+            }
+
+            // Check if there is a continuous increase in count of unprocessed changes.
+            if (metrics[0].UnprocessedChangeCount > 0)
```

Comment:
Why only the first in the array? Is that going to be the first (earliest) or last (latest) metrics gathered?

Reply:
It is going to be the earliest in the window. Most of the logic is copied from the [Azure Queue listener](https://github.com/Azure/azure-webjobs-sdk/blob/919d2d63419691166a6e36f1c1785f4c2c31aeb3/src/Microsoft.Azure.WebJobs.Extensions.Storage/Queues/Listeners/QueueListener.cs#L519). The check against 0 is not very critical. It only ensures that we scale for metrics [10, 20, 30] and not for [0, 10, 20] as in a way, that means the SQL table was not active for the entirety of the window.

Reply:
I'd probably take their IsTrueForLastN over too - that seems useful and will make this easier to understand.

Reply:
The queue listener needs to loop 4 times so there it makes sense to move the iteration in its own method. The next commit is going to have a single loop anyhow. I am also finding the `if (metrics[0].UnprocessedChangeCount > 0)` to be arbitrary, so that will also go away.

Another problem I see with the queue-listener logic is that if the function app is lightly loaded, since we are doing comparison only among 5 samples, there is possibility that they turn out to be all in an increasing order e.g. [0, 2, 4, 5, 9]. Since the max capacity per worker is 1000, this should not cause a scale out. Similarly, if we find the metric samples like [1900, 1988, 1955, 1944, 1800] with two workers, this should not cause as scale in, since clearly 1 worker is not going to handle these many requests. I think most of the logic should hover around current number of changes, and the trend of increase/decrease should be a secondary consideration.

Reply:
For trend, a better approach should be to extrapolate and if we find the number of pending changes will go beyond the total capacity of existing workers, then we should scale out. Same logic can determine when one should scale-in. Of course, this will require a couple of unit tests to ensure there are no bugs.

Reply:
This is done from my end. The unit tests will be added to verify there are no issues in the next PR.